### ISSUE TITLE
multi-cluster support with envtest

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra-operator#35]
We use the `envtest` package for controller integration tests. See this [section](https://book.kubebuilder.io/cronjob-tutorial/writing-tests.html) from the kubebuilder book for background. `envtest` deploys the api server and etcd locally. It would be nice if we could figure out how to adapt this for multi-cluster scenarios

I came up short with my initial investigation. In test I started out by trying to deploy a second server using the defaults just to see what problems I would hit. My setup code for one server looks like this:

```go
        var testEnv *envtest.Environment

	testEnv = &envtest.Environment{
		CRDDirectoryPaths: []string{
			filepath.Join("..", "config", "crd", "bases"),
			filepath.Join("..", "config", "cass-operator", "crd", "bases")},
	}

	cfg, err = testEnv.Start()
	require.NoError(err, "failed to start test environment")

	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
		Scheme: scheme.Scheme,
	})
	require.NoError(err, "failed to create controller-runtime manager")
```

I just duplicated this code to configure and start the second server. The second server failed to start due to a port conflict for the insecure port on which the api server listens for client requests. I got past the error with changing my configuration as follows:

```go
	testEnv2 := &envtest.Environment{
		CRDDirectoryPaths: []string{
			filepath.Join("..", "config", "crd", "bases"),
			filepath.Join("..", "config", "cass-operator", "crd", "bases")},
		KubeAPIServerFlags: []string{
			"--insecure-port", "8081"},
                 AttachControlPlaneOutput: true,
	}
```

Next I got a file systems permissions error because the server could not write certs to a directory. I fixed this with:

```go
	testEnv2 := &envtest.Environment{
		CRDDirectoryPaths: []string{
			filepath.Join("..", "config", "crd", "bases"),
			filepath.Join("..", "config", "cass-operator", "crd", "bases")},
		KubeAPIServerFlags: []string{
			"--insecure-port", "8081",
			"--cert-dir", certDir},
                 AttachControlPlaneOutput: true,
	}
```

That is about as much progress that I made so far. `testenv` applies several default values  for various options to both the api server and etcd when no options are specified. These defaults are tailored for a local testing environment. When any options are specified for `KubeAPIServerFlags`, it looks like `envtest` applies none of its defaults.

With the above configuration I get this error message in the start up logs of the second server:

```
Error: [--etcd-servers must be specified, specified --service-cluster-ip-range is too large; for 32-bit addresses, the mask must be >= 12]
```

Then I added `"--etcd-servers", "http://127.0.0.1:56750"` but the server start up still fails with this error:

`E0709 17:11:41.828445   45837 controller.go:184] Get "https://[::1]:6443/api/v1/namespaces/default/endpoints/kubernetes": dial tcp [::1]:6443: connect: connection refused`.

┆Issue is synchronized with this [Jiraserver Task](https://k8ssandra.atlassian.net/browse/K8SSAND-690) by [Unito](https://www.unito.io)
┆Epic: CI/CD Environment Bootstrap
┆Fix Versions: k8ssandra-2.0.0-alpha
┆Issue Number: K8SSAND-690
┆Priority: Medium

**Comment by burmanm** [comment]
Do we need multiple envtests in the same running machine? Or could we run them in two different docker containers? Or even possibly with two different IPs (127.0.0.2 / 127.0.0.3 for example).

**Comment by jsanda** [comment]
I thought about running them in docker containers. This would allow us to run with the default configuration for each server. For envtest servers we don't need to worry about IP connectivity between clusters because no pods are actually deployed. Running them in containers is certainly something to consider; however, we should be able to run multiple instances in parallel.

I suspect we will still run into some of the types of challenges I described even with using different IPs. When default settings are used, envtest calls the [Suggest](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.8.3/pkg/internal/testing/integration/addr/manager.go#L65) function to find a port that is not in use. This is used in create the urls for both the api server and for etcd.

While I am eager to get this sorted out, I need to shift my focus to how handle setup for multi-cluster e2e tests.

**Comment by jsanda** [comment]
This was fixed by #51.
